### PR TITLE
[SPARK-10530] [CORE] Kill other task attempts when one taskattempt belonging 

### PR DESCRIPTION
**Pull request by zjffdu**
SPARK-10530 Kill other task attempts when one taskattempt belonging the same task is succeeded in speculation

Verify it manually. Unit test will be added in the followup commit. Initial review on the approach is very appreciated.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 1]
```diff
@@ -628,6 +628,9 @@ private[spark] class TaskSetManager(
     // here "result.value()" just returns the value and won't block other threads.
     sched.dagScheduler.taskEnded(
       tasks(index), Success, result.value(), result.accumUpdates, info, result.metrics)
+    // kill other attempts of this task when speculation is enabled.
+    taskAttempts(index).filter(_.taskId != tid).foreach(e=>sched.killSpeculatedTask(e.taskId))
```
Comment (dragos) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: It'd be slightly better to rewrite this to 

```
for (e <- taskAttempts(index) if e.taskId != tid) 
  sched.killSpeculatedTast..
```

The Scala compiler would translate the if-guard to a `withFilter` call, which is more efficient in that it won't create a new collection (which filter does).

**Comment by dragos** [comment]
Please modify the PR title to match the guidelines (`[SPARK-XXXXX][Component] <title>`)

**Comment by zjffdu** [comment]
Thanks for the tips @dragos 
Push another commit with the changes

**Comment by tgravescs** [comment]
Jenkins, test this please

**Comment by tgravescs** [comment]
what testing have you done on this?  I think the DAGSchedulerSuite will need to be updated for the TAskScheduler api.  It would be nice if we can add a test for this scenario as well.

**Comment by zjffdu** [comment]
@tgravescs sorry for late response, I will add unit test in the next few days.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 2]
```diff
@@ -628,6 +628,11 @@ private[spark] class TaskSetManager(
     // here "result.value()" just returns the value and won't block other threads.
     sched.dagScheduler.taskEnded(
       tasks(index), Success, result.value(), result.accumUpdates, info, result.metrics)
+    // kill other attempts of this task when speculation is enabled.
```
Comment (chenghao-intel) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I am not so sure about this part, but will that be better if we moved the code after we confirmed the task succeeded?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 3]
```diff
@@ -220,6 +220,15 @@ private[spark] class TaskSchedulerImpl(
     }
   }
 
+  def killSpeculatedTask(taskId:Long) : Unit = synchronized {
```
Comment (chenghao-intel) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Nit: space after `:`

**Comment by tgravescs** [comment]
@zjffdu  any update on this?

**Comment by kayousterhout** [comment]
Can you close this pull request? It has been fixed by #11996

**Pull request closed without merging**